airflow-core/src/airflow/migrations/versions/0074_3_1_0_make_bundle_name_not_nullable.py

```diff
@@ -0,0 +1,114 @@
+branch_labels = None
+depends_on = None
+airflow_version = "3.1.0"
+
+
+def upgrade():
+    """Make bundle_name not nullable."""
+    dialect_name = op.get_bind().dialect.name
+    if dialect_name == "postgresql":
+        op.execute("""
+                    INSERT INTO dag_bundle (name) VALUES
+                      ('example_dags'),
+                      ('dags-folder')
+                    ON CONFLICT (name) DO NOTHING;
+                    """)
+    if dialect_name == "mysql":
+        op.execute("""
+                    INSERT IGNORE INTO dag_bundle (name) VALUES
+                      ('example_dags'),
+                      ('dags-folder');
+                    """)
+    if dialect_name == "sqlite":
+        op.execute("""
+                    INSERT OR IGNORE INTO dag_bundle (name) VALUES
+                      ('example_dags'),
+                      ('dags-folder');
+                    """)
+
+    conn = op.get_bind()
+    with op.batch_alter_table("dag", schema=None) as batch_op:
+        conn.execute(
+            text(
+                """
+                UPDATE dag
+                SET bundle_name =
+                    CASE
+                        WHEN fileloc LIKE '%/airflow/example_dags/%' THEN 'example_dags'
+                        ELSE 'dags-folder'
+                    END
+                WHERE bundle_name IS NULL
+                """
+            )
+        )
+        # drop the foreign key temporarily and recreate it once both columns are changed
+        batch_op.drop_constraint(batch_op.f("dag_bundle_name_fkey"), type_="foreignkey")
+        batch_op.alter_column("bundle_name", nullable=False, existing_type=sa.String(length=250))
+
+    with op.batch_alter_table("dag_bundle", schema=None) as batch_op:
+        batch_op.alter_column("name", nullable=False, existing_type=sa.String(length=250))
+
+    with op.batch_alter_table("dag", schema=None) as batch_op:
+        batch_op.create_foreign_key(
+            batch_op.f("dag_bundle_name_fkey"), "dag_bundle", ["bundle_name"], ["name"]
+        )
+
+
+def downgrade():
+    """Make bundle_name nullable."""
+    with op.batch_alter_table("dag", schema=None) as batch_op:
+        batch_op.drop_constraint(batch_op.f("dag_bundle_name_fkey"), type_="foreignkey")
+
+        existing_type_kwargs = {}
+        if op.get_bind().dialect.name == "mysql":
+            existing_type_kwargs["collation"] = "utf8mb3_bin"
+        batch_op.alter_column(
+            "bundle_name", nullable=True, existing_type=sa.String(length=250, **existing_type_kwargs)
+        )
+    if op.get_bind().dialect.name == "mysql":
+        # Ensures the collation and charset are the same before creating the foreign key
+        op.execute(
+            "ALTER TABLE dag_bundle MODIFY name VARCHAR(250) CHARACTER SET utf8mb3 COLLATE utf8mb3_bin"
+        )
+        op.execute(
+            "ALTER TABLE dag MODIFY bundle_name VARCHAR(250) CHARACTER SET utf8mb3 COLLATE utf8mb3_bin"
+        )
```

Comment:
Using the `StringID` from `airflow.migrations.db_types` saves us from doing this

Reply:
I will give it a try, but I'm also a little bit confused about StringID - 
`StringID` from `airflow.migrations.db_types` is nothing but an alias of `sa.String`, so seems like it's not doing anything extra.
While there's another `StringID` from `airflow.models.base` that adds `{"collation": "utf8mb3_bin"}`. Do you mean using this one?

Reply:
Thanks! Good to know this. Migration updated.